pallets/market/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+    DisallowedProduct(AssetId),
+    Expirable(BlockNumber),
+}
+
+#[derive(RuntimeDebug, Encode, Decode, TypeInfo, Clone, PartialEq, Eq)]
+pub struct Product<BlockNumber, AssetId> {
+    price_asset: AssetId,
+    price: Balance,
+    extensions: Vec<Extension<BlockNumber, AssetId>>,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+
+    use super::*;
+
+    /// The current storage version.
+    const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(_);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config + common::Config {
+        /// Event type of this pallet.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        type AssetInfoProvider: common::AssetInfoProvider<
+            AssetIdOf<Self>,
+            AccountIdOf<Self>,
+            common::AssetSymbol,
+            common::AssetName,
+            common::BalancePrecision,
+            common::ContentSource,
+            common::Description,
+        >;
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn product)]
+    pub type Products<T: Config> = StorageMap<
+        _,
+        Identity,
+        AssetIdOf<T>,
+        Product<BlockNumberFor<T>, AssetIdOf<T>>,
+        OptionQuery,
+    >;
+
+    #[pallet::storage]
+    #[pallet::getter(fn expiration)]
+    pub type Expirations<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        BlockNumberFor<T>,
+        Identity,
+        (T::AccountId, AssetIdOf<T>),
+        Balance,
+        ValueQuery,
+    >;
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+        fn on_initialize(now: T::BlockNumber) -> Weight {
+            let mut processed_expirations = 0;
+            for ((account_id, product_id), value_to_burn) in Expirations::<T>::drain_prefix(now) {
+                processed_expirations += 1;
+                if let Err(err) =
+                    T::AssetManager::burn_unchecked(&product_id, &account_id, value_to_burn)
+                {
+                    frame_support::log::warn!(
+                        "Unable to burn expired product, product_id: {:?}, account_id {:?}, amount: {}, err: {:?}",
+                        product_id,
+                        account_id,
+                        value_to_burn,
+                        err
+                    );
+                }
+            }
+            <T as Config>::WeightInfo::on_initialize(processed_expirations)
+        }
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        ProductRegistered { asset_id: AssetIdOf<T> },
```

Comment:
I think we need `Buy` event as well. It can make easier some search in indexer when it is necessary.